File: src/economy/woodcutterPolicy.ts
export const WOODCUTTER_TIMBER_RESERVE_MIN = 0;
export const WOODCUTTER_TIMBER_RESERVE_MAX = 240;

export const WOODCUTTER_TIMBER_RESERVE_PRESETS = [
  { reserve: 0, label: 'Burn freely' },
  { reserve: 40, label: 'Cottage plan' },
  { reserve: 100, label: 'Workshop plan' },
  { reserve: 200, label: 'Civic plan' },
] as const;

const STOCK_EPSILON = 1e-6;

export function normalizeWoodcutterTimberReserve(reserve: number): number {
  if (!Number.isFinite(reserve)) return WOODCUTTER_TIMBER_RESERVE_MIN;
  return Math.max(
    WOODCUTTER_TIMBER_RESERVE_MIN,
    Math.min(WOODCUTTER_TIMBER_RESERVE_MAX, Math.round(reserve)),
  );
}

export function woodcutterCanProcess(
  availableUnreservedTimber: number,
  timberReserve: number,
  timberNeeded: number,
): boolean {
  if (!Number.isFinite(availableUnreservedTimber) || !Number.isFinite(timberNeeded)) {
    return false;
  }
  return availableUnreservedTimber + STOCK_EPSILON
    >= normalizeWoodcutterTimberReserve(timberReserve) + Math.max(0, timberNeeded);
}

export function timberAboveWoodcutterReserve(
  availableUnreservedTimber: number,
  timberReserve: number,
): number {
  return Math.max(
    0,
    availableUnreservedTimber - normalizeWoodcutterTimberReserve(timberReserve),
  );
}

File: server/src/woodcutter_policy.rs
pub const WOODCUTTER_TIMBER_RESERVE_MIN: f64 = 0.0;
pub const WOODCUTTER_TIMBER_RESERVE_MAX: f64 = 240.0;

pub fn normalize_woodcutter_timber_reserve(reserve: f64) -> f64 {
    if !reserve.is_finite() {
        return WOODCUTTER_TIMBER_RESERVE_MIN;
    }
    reserve
        .round()
        .clamp(WOODCUTTER_TIMBER_RESERVE_MIN, WOODCUTTER_TIMBER_RESERVE_MAX)
}

#[cfg(test)]
mod tests {
    use super::{normalize_woodcutter_timber_reserve, WOODCUTTER_TIMBER_RESERVE_MAX};

    #[test]
    fn reserve_is_rounded_and_bounded() {
        assert_eq!(normalize_woodcutter_timber_reserve(-10.0), 0.0);
        assert_eq!(normalize_woodcutter_timber_reserve(39.6), 40.0);
        assert_eq!(
            normalize_woodcutter_timber_reserve(10_000.0),
            WOODCUTTER_TIMBER_RESERVE_MAX
        );
    }
}
